File: 01-programming-languages/Kotlin/01-fundamentals/variables.kt
fun main() {
    // Immutable variables (val) - cannot be reassigned
    val name = "Kotlin"
    val version = 1.9
    val isAwesome = true
    println("Language: $name, Version: $version, Awesome: $isAwesome")

    // Mutable variables (var) - can be reassigned
    var counter = 0
    counter = 1
    counter += 10
    println("Counter: $counter")

    // Type inference
    val inferredString = "Hello, World!"
    val inferredInt = 42
    val inferredDouble = 3.14
    val inferredBoolean = false
    println("$inferredString, $inferredInt, $inferredDouble, $inferredBoolean")

    // Explicit types
    val explicitString: String = "Explicitly typed"
    val explicitInt: Int = 100
    val explicitLong: Long = 1_000_000L
    val explicitFloat: Float = 2.5f
    val explicitDouble: Double = 2.5
    val explicitByte: Byte = 127
    val explicitShort: Short = 32767
    val explicitChar: Char = 'A'
    println("Types: $explicitString, $explicitInt, $explicitLong, $explicitFloat, $explicitDouble")

    // String templates
    val language = "Kotlin"
    val year = 2011
    println("$language was released in $year")
    println("Length of language: ${language.length}")
    println("Result: ${if (year > 2010) "Modern" else "Legacy"}")

    // Multi-line strings
    val multiLine = """
        This is a multi-line string.
        It can span multiple lines.
        Trim margin: ${multiLine.trimMargin()}
    """.trimMargin()
    println(multiLine)

    // Constants (compile-time)
    const val MAX_SIZE = 100
    const val APP_NAME = "MyApp"
    println("Max size: $MAX_SIZE, App: $APP_NAME")

    // Arrays and Lists
    val list = listOf(1, 2, 3, 4, 5)
    val mutableList = mutableListOf(1, 2, 3)
    mutableList.add(4)
    println("List: $list, Mutable: $mutableList")

    // Maps
    val map = mapOf("a" to 1, "b" to 2, "c" to 3)
    val mutableMap = mutableMapOf("x" to 10)
    mutableMap["y"] = 20
    println("Map: $map, Mutable: $mutableMap")

    // Destructuring
    val (a, b, c) = listOf(10, 20, 30)
    println("Destructured: $a, $b, $c")
}

File: 01-programming-languages/Rust/01-fundamentals/variables.rs
fn main() {
    // Immutable variables (let)
    let name = "Rust";
    let version = 1.75;
    let is_awesome = true;
    println!("Language: {}, Version: {}, Awesome: {}", name, version, is_awesome);

    // Mutable variables (mut)
    let mut counter = 0;
    counter = 1;
    counter += 10;
    println!("Counter: {}", counter);

    // Type inference
    let inferred_int = 42;
    let inferred_float = 3.14;
    let inferred_bool = false;
    let inferred_char = 'A';
    println!("{}, {}, {}, {}", inferred_int, inferred_float, inferred_bool, inferred_char);

    // Explicit types
    let explicit_int: i32 = 100;
    let explicit_long: i64 = 1_000_000;
    let explicit_float: f32 = 2.5;
    let explicit_double: f64 = 2.5;
    let explicit_byte: u8 = 255;
    let explicit_short: u16 = 32767;
    println!("{}, {}, {}, {}, {}, {}", explicit_int, explicit_long, explicit_float, explicit_double, explicit_byte, explicit_short);

    // String formatting
    let language = "Rust";
    let year = 2015;
    println!("{} was released in {}", language, year);
    println!("Length of language: {}", language.len());
    println!("Result: {}", if year > 2010 { "Modern" } else { "Legacy" });

    // Multi-line strings
    let multi_line = "
        This is a multi-line string.
        It can span multiple lines.
    ";
    println!("{}", multi_line);

    // Constants
    const MAX_SIZE: u32 = 100;
    const APP_NAME: &str = "MyApp";
    println!("Max size: {}, App: {}", MAX_SIZE, APP_NAME);

    // Arrays
    let array = [1, 2, 3, 4, 5];
    let mutable_array = [1, 2, 3];
    println!("Array: {:?}, Mutable: {:?}", array, mutable_array);

    // Vectors
    let vector = vec![1, 2, 3, 4, 5];
    let mut mutable_vector = vec![1, 2, 3];
    mutable_vector.push(4);
    println!("Vector: {:?}, Mutable: {:?}", vector, mutable_vector);

    // Tuples
    let tuple = (1, 2.0, "hello", true);
    let (a, b, c, d) = tuple;
    println!("Tuple: {}, {}, {}, {}", a, b, c, d);

    // Destructuring
    let (x, y, z) = (10, 20, 30);
    println!("Destructured: {}, {}, {}", x, y, z);

    // Shadowing
    let shadow = 5;
    let shadow = shadow + 1;
    let shadow = shadow * 2;
    println!("Shadow: {}", shadow);

    // Type shadowing
    let spaces = "   ";
    let spaces = spaces.len();
    println!("Spaces: {}", spaces);
}
